bricklayers.py

```diff
@@ -1,143 +1,282 @@
+                        modified_lines.append(f"G1 Z{adjusted_z:.3f} F1200 ; Z shift for internal perimeter\n")
+                        self.shifted_blocks += 1
+                        
+                        # Adjust extrusion
+                        e_match = re.search(r'E([-\d.]+)', line)
+                        if e_match:
+                            e_value = float(e_match.group(1))
+                            if self.current_layer == 0:
+                                new_e_value = e_value * 1.5
+                                comment = "first layer"
+                            elif is_last_layer:
+                                new_e_value = e_value * 0.5
+                                comment = "last layer"
+                            else:
+                                new_e_value = e_value * self.extrusion_multiplier
+                                comment = "internal perimeter"
+                            
+                            line = re.sub(r'E[-\d.]+', f'E{new_e_value:.5f}', line.strip())
+                            line += f" ; Adjusted E for {comment}\n"
+                    else:
+                        modified_lines.append(f"G1 Z{self.current_z:.3f} F1200 ; Reset Z for even block\n")
+            
+            elif "F" in line and not "E" in line and self.inside_perimeter_block:
+                modified_lines.append(f"G1 Z{self.current_z:.3f} F1200 ; Reset Z after internal perimeter\n")
+                self.inside_perimeter_block = False
+
+        modified_lines.append(line)
+        return modified_lines
+
+    def process_gcode(self, input_file, is_bgcode=False):
+        """Process G-code file using buffered reading and writing"""
+        # Set up logging
+        script_dir = os.path.dirname(os.path.abspath(__file__))
+        log_file_path = os.path.join(script_dir, "z_shift_log.txt")
+        logging.basicConfig(
+            filename=log_file_path,
+            filemode="w",
+            level=logging.INFO,
+            format="%(asctime)s - %(message)s"
+        )
+        
+        logging.info("Starting G-code processing")
+        logging.info(f"Input file: {input_file}")
+        logging.info(f"Z-shift: {self.z_shift} mm, Layer height: {self.layer_height} mm")
+
+        # Handle bgcode conversion if necessary
+        input_path = Path(input_file)
+        if is_bgcode:
+            if input_path.suffix != '.bgcode':
+                input_path = input_path.rename(str(input_path) + '.bgcode')
+            if os.system('bgcode ' + str(input_path)) == 0:
+                input_path = input_path.parent / (input_path.stem + '.gcode')
+
+        # First pass to count layers and detect printer type
+        with open(input_path, 'r') as infile:
+            content = infile.read()
+            lines = content.splitlines(True)
+            self.printer_type = self.detect_printer_type(lines)
+            self.total_layers = sum(1 for line in lines if line.startswith("G1 Z"))
```

Comment:
```suggestion
            lines = infile.readlines()
            self.printer_type = self.detect_printer_type(lines)
            self.total_layers = sum(line.startswith("G1 Z") for line in lines)
```
just a little suggestion on simplifying a little